jobs/inspect_archives.py

```diff
@@ -0,0 +1,221 @@
+    return id_, version
+
+
+def process_linux_archive(client, record):
+    # Download archive
+    url = record["download"]["url"]
+
+    has_sysaddons = int(record["target"]["version"].split(".", 1)[0]) > 50
+    systemaddons = [] if has_sysaddons else None
+
+    updated = record.copy()
+
+    archive = cached_download(url)
+
+    size = os.path.getsize(archive)
+    if url.endswith("bz2"):
+        mode = "r:bz2"
+        mimetype = "application/x-bzip2"
+    else:
+        mode = "r:gz"
+        mimetype = "application/gzip"
+
+    updated["download"]["size"] = size
+    updated["download"]["mimetype"] = mimetype
+
+    try:
+        logger.info("Extract from %s" % url)
+        tar = tarfile.open(archive, mode)
+    except tarfile.ReadError:
+        logger.error("Could not read archive %s." % archive)
+        # Will retry download next run.
+        os.remove(archive)
+        return
+    try:
+        for tarinfo in tar:
+            filename = tarinfo.name
+
+            # Inspect metadata.ini
+            if filename.endswith("application.ini"):
+                logger.info("Read %s" % filename)
+                f = tar.extractfile(tarinfo)
+                ini_content = f.read()
+                metadata = extract_application_metadata(ini_content)
+                updated["build"] = {**(updated.get("build") or {}), **metadata["build"]}
+                updated["source"] = {**(updated.get("source") or {}), **metadata["source"]}
+
+                if not has_sysaddons:
+                    break
+
+            # Inspect system addons
+            if has_sysaddons and re.match(r".+/browser/features/.+\.xpi$", filename):
+
+                # XXX: Apparently Python can't unzip XPI files oO
+                # zipfile.BadZipFile: Bad magic number for central directory
+                with tempfile.TemporaryDirectory() as tmpdirname:
+                    logger.info("Introspect system-addon %s" % filename)
+                    tar.extractall(path=tmpdirname, members=[tarinfo])
+                    xpipath = os.path.join(tmpdirname, filename)
+
+                    logger.debug("Extract install manifest from %s" % xpipath)
+                    subprocess.run(["unzip", "-o", xpipath, "install.rdf", "-d", tmpdirname])
+                    xmlfile = os.path.join(tmpdirname, 'install.rdf')
+
+                    logger.debug("Read version from %s" % xmlfile)
+                    xml_content = open(xmlfile).read()
+                    id_, version = extract_systemaddon_metadata(xml_content)
+                    systemaddons.append({"id": id_, "builtin": version})
+
+        tar.close()
+    except:
+        logger.exception("Could not introspect archive %s" % url)
+
+    updated["systemaddons"] = systemaddons
+
+    has_changed = json.dumps(record, sort_keys=True) != json.dumps(updated, sort_keys=True)
+    if has_changed:
+        logger.info("Update metadata of %s" % updated)
+        # XXX why 412 here? why safe=False?
```

Comment:
You will get a 412 if the updated record contains a last_modified and if it doesn't match the if-match header of the server.